build: make Lua detection more friendly?

qmega opened the pull request:
Picking up from #8355.

The upside is you can use `./waf configure --lua=52` regardless of how your distro packages Lua versions. Also threw in a little over-engineering to make the used Lua version a bit clearer in logs ("List of enabled features" just showed "52", will now show "lua52").

The downside is losing the ability to pick between the different packaging variants if you somehow end up with more than one for the same version on your machine, and that logs will no long differentiate between said variants. I don't see much harm there, but admittedly the benefit of this patch is similarly small.

The first commit at least should be a simple win, though.

avih commented:
Not sure how relevant it is, but previously you could select a specific package in case you had more than one on your system, but now you can only specify the "category" and the chosen package will depend on the search order (if more than one matches).

It does seem more convenient to use though if you don't have more than one in each "category".

qmega commented:
> Not sure how relevant it is, but previously you could select a specific package in case you had more than one on your system, but now you can only specify the "category" and the chosen package will depend on the search order (if more than one matches).

Yeah, that's the first downside I mentioned. I thought it would be very unlikely for someone to end up with more than one pkg-config for the same Lua version *and* have a reason to prefer one over another. I suppose it still isn't great to remove an technically existing capability for limited benefit, though.

We could allow specifying the versions by their pkg-config names, which IMO makes more sense than the distro-specific naming. But perhaps I'm over-reacting after all. I basically just didn't like the idea of having to use "51obsd" on Arch or "52arch" on OpenBSD.

In any case, allowing 51/52 to try them all still seems nice.

> Also, nit, if you already use the name as "51" and "52", wouldn't it be more consistent to also use the name "jit" instead of "luajit"?

My feeling is that "jit" isn't a "version" of lua and `--lua=jit` would feel weird to me. Maybe that's just me, though. I'm now leaning toward giving them all multiple aliases, though, so maybe we just allow both.

> Another thing we should keep in mind is that this change will break build scripts which currently specify a package name, because the previous names are now gone.

Yeah. I actually had a change queued up for AUR/mpv-git if this PR went through. Keeping the old names as aliases is probably better.

qmega commented:
Updated. Now accepts lua51, lua52, luajit, 51, 52, jit, and all the old distro-specific names as well.

Only works on Python 3 due to the rest unpacking (`*_`). We've dropped Python 2, right?

avih commented:
Rethinking about it, I'm not sure how much value it adds.

Do correct me if I'm wrong, but the only use case which the updated PR improves is when the user has more than one of luajit/51/52, and they want to choose the version which has lower priority in our search order but without specifying the explicit package name, right?

I.e. if they only have one version then no change, if they have more than one but their preference is the same as our search order then no change.

So, while the last update seems completely backward compatible, ultimately the specific use case which becomes more convenient is, in my subjective opinion, not worth the additional code, especially because lua detection was already messy code, and additional code/complexity for minor convenience is not a great bargain IMHO.

avih commented:
Actually, it also doesn't change when they want luajit, because it has only one variant anyway.

So in fact, the single use case it improves is that they want lua 5.1 but also have jit and/or 5.2 installed, and now they can specify "51" instead of one of the explicit 5.1 package names, yes?

qmega commented:
> So in fact, the single use case it improves is that they want lua 5.1 but also have jit and/or 5.2 installed, and now they can specify "51" instead of one of the explicit 5.1 package names, yes?

Yeah, I guess so. It also makes the "List of enabled features" in logs show "lua52"/"lua51" instead of just "52"/"51".

You're probably right that this is over-engineered and not worth it. I've certainly spent more time on it than I intended to already, and I don't think it's worth any more.

qmega commented:
Dropped the complicated stuff. I still think 6c568aca2a1ddcdb4f30fd92fa1fcd3e7efe42e6 should be done. Just added a revert of the redundant change that started me down this path.

sfan5 commented:
I liked this PR in it's very first version but the current (reduced) changes should also be merged.

qmega commented:
Well, I don't know how the decision of which gets merged would be made, but all the commits are still available and I have no special attachment to any one version over the others, FWIW. Feel free to take any of them. If you'd like me to push a particular version to this branch for whatever reason let me know.

avih commented:
FWIW, I don't have a strict objection to any version. It does seem to me borderline not-worth-it , but if others want it, no objection here.
